### ISSUE TITLE
Sending Pre-Compressed Buffers?

### ISSUE THREAD

**Issue by ronag** [uNetworking/uWebSockets.js#1076]
Would it be possible to send pre-compressed data? i.e. if one caches messages one could cache the compressed version in order to avoid compressing during every send?

e.g.

Instead of: 

```js
cache.set(key, buf)
// ...
for (const socket of sockets) (  
  socket.send(cache.get(key)) // compressed by uws
}
```

```js
cache.set(key, zlib.deflateSync(buf))
// ...
for (const socket of sockets) (
  socket.send(cache.get(key))
}
```

This would enable 2 things:

- Re-use compression
- Move compression to other thread

**Comment by ronag** [comment]
If this is possible to implement I would consider funding the work.

**Comment by uNetworkingAB** [comment]
Yes it is possible. It was the behaviour in v0.14. But doing more investigation I found that, this was very poor for both compressive performance, and for CPU performance. 


My first idea was that sharing a compressor would be best, but the cost of compressing is directly (linearly) correlated with the difference from what you have in your sliding window and what you are compressing. 

This means that dedicated (per connection) compression is both faster and obviously compresses better. 

If the difference is minimal, compression costs almost nothing. If the difference is total, it's slow as a snail. 

Most who use compression prefers the dedicated compression. So the shared pub sub compression was removed (because of complexity and added work of keeping multiple variants of the outgoing data).

Adding something like a shared compressor (only) for pub sub is most likely doable

**Comment by klauss194** [comment]
+1

**Comment by uNetworkingAB** [comment]
You can only pre-compress if you use shared compression (no sliding window). But compressing without sliding window is extremely slow compared to having a sliding window because compression cost is directly correlated with the difference "damage" from previous message. So with shared compression, the difference is always maximal. So most users simply pick dedicated compressor.

That's why the feature was removed. It's a known trick but nobody really wants it.
